Client/Frontend/Settings/SearchCustomEngineViewController.swift

```diff
@@ -0,0 +1,727 @@
+              let faviconImage = faviconImage,
+              let hostURLString = host?.absoluteString else {
+            let alert = ThirdPartySearchAlerts.failedToAddThirdPartySearch()
+            present(alert, animated: true, completion: nil)
+            return
+        }
+        
+        while referenceURLString.hasPrefix("/") {
+            referenceURLString.remove(at: referenceURLString.startIndex)
+        }
+        
+        let constructedReferenceURLString = "\(hostURLString)/\(referenceURLString)"
+
+        if referenceURL.host == nil, let constructedReferenceURL = URL(string: constructedReferenceURLString) {
+            referenceURL = constructedReferenceURL
+        }
+            
+        downloadOpenSearchXML(referenceURL, referenceURL: referenceURLString, title: title, iconImage: faviconImage)
+    }
+    
+    func downloadOpenSearchXML(_ url: URL, referenceURL: String, title: String, iconImage: UIImage) {
+        changeAddButton(for: .loading)
+        view.endEditing(true)
+        
+        NetworkManager().downloadResource(with: url).uponQueue(.main) { [weak self] response in
+            guard let self = self else { return }
+            
+            if let openSearchEngine = OpenSearchParser(pluginMode: true).parse(response.data, referenceURL: referenceURL, image: iconImage, isCustomEngine: true) {
+                self.addSearchEngine(openSearchEngine)
+            } else {
+                let alert = ThirdPartySearchAlerts.failedToAddThirdPartySearch()
+                
+                self.present(alert, animated: true) {
+                    self.changeAddButton(for: .disabled)
+                }
+            }
+        }
+    }
+    
+    func addSearchEngine(_ engine: OpenSearchEngine) {
+        let alert = ThirdPartySearchAlerts.addThirdPartySearchEngine(engine) { [weak self] alertAction in
+            guard let self = self else { return }
+            
+            if alertAction.style == .cancel {
+                self.changeAddButton(for: .enabled)
+                return
+            }
+            
+            do {
+                try self.profile.searchEngines.addSearchEngine(engine)
+                self.cancel()
+            } catch {
+                self.handleError(error: SearchEngineError.failedToSave)
+                
+                self.changeAddButton(for: .disabled)
+            }
+        }
+
+        self.present(alert, animated: true, completion: {})
+    }
+}
+
+// MARK: Auto Add Meta Data
+
+extension SearchCustomEngineViewController {
+    
+    func checkSupportAutoAddSearchEngine() {
+        guard let openSearchEngine = openSearchEngine else {
+            changeAddButton(for: .disabled)
+            checkManualAddExists()
+
+            faviconImage = nil
+            
+            return
+        }
+        
+        let matches = profile.searchEngines.orderedEngines.filter {$0.referenceURL == openSearchEngine.reference}
```

Comment:
Change this to use  `.contains(where: { ... })` or `.first(where: {  })` since you are not actually relying on the filtered data set

Reply:
Good catch, Contains makes more sense and faster.